hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/opt/physical/JoinHashPhysicalRel.java

```diff
@@ -58,9 +58,34 @@ public Join copy(
         return new JoinHashPhysicalRel(getCluster(), traitSet, left, right, conditionExpr, joinType);
     }
 
+    /**
+     * Cost calculation of Hash Join relation. It does not rely on children cost.
+     * <p>
+     * Hash Join algorithm is a more advanced join algorithm, which builds a hash table for the left
+     * row set, and then compare each row from the right side. Cost estimation is the following: <ol>
+     * <li> Produced row count is L * R assuming the join selectivity is 1.
+     * <li> Processed row count is L + R because we traverse both sides once per JOIN.
+     * <li> CPU is L * (hash table build cost) + R * (row comparison cost). </ol>
+     * <p>
+     * A perfect estimation must also include memory (occupied by the hash table) and IO costs,
+     * as well as a selectivity for the right row set.
+     */
     @Override
     @Nullable
     public RelOptCost computeSelfCost(RelOptPlanner planner, RelMetadataQuery mq) {
-        return super.computeSelfCost(planner, mq).multiplyBy(COST_FACTOR);
+        double leftRowCount = mq.getRowCount(getLeft());
+        double rightRowCount = mq.getRowCount(getRight());
+
+        Double selectivity = mq.getSelectivity(this, condition);
+        if (selectivity == null) {
+            selectivity = 1.;
+        }
+
+        // TODO: introduce selectivity estimator, but ATM we taking the worst case scenario : selectivity = 1.0.
+        double producedRowCount = mq.getRowCount(this);
+        double cpu = leftRowCount * Cost.HASH_JOIN_MULTIPLIER
+                + rightRowCount * selectivity * Cost.JOIN_ROW_CMP_MULTIPLIER;
```

Comment:
Join selectivity only affects the number of produced rows in hash join. You compute the predicate for unmatched rows as well.
```suggestion
                + rightRowCount * Cost.JOIN_ROW_CMP_MULTIPLIER;
```

Reply:
True.